File: app/src/main/java/com/azreashade/shadowpurge/AppRepository.java
package com.azreashade.shadowpurge;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class AppRepository {
    private final PackageManager pm;

    public AppRepository(Context ctx) {
        pm = ctx.getPackageManager();
    }

    public List<AppInfo> getUserApps() {
        List<ApplicationInfo> list = pm.getInstalledApplications(PackageManager.GET_META_DATA);
        List<AppInfo> out = new ArrayList<>();
        for (ApplicationInfo ai : list) {
            boolean isSystem = (ai.flags & ApplicationInfo.FLAG_SYSTEM) != 0;
            if (!isSystem) {
                out.add(new AppInfo(pm.getApplicationLabel(ai).toString(), ai.packageName, false));
            }
        }
        sort(out);
        return out;
    }

    public List<AppInfo> getSystemApps() {
        List<ApplicationInfo> list = pm.getInstalledApplications(PackageManager.GET_META_DATA);
        List<AppInfo> out = new ArrayList<>();
        for (ApplicationInfo ai : list) {
            boolean isSystem = (ai.flags & ApplicationInfo.FLAG_SYSTEM) != 0;
            if (isSystem) {
                out.add(new AppInfo(pm.getApplicationLabel(ai).toString(), ai.packageName, true));
            }
        }
        sort(out);
        return out;
    }

    private void sort(List<AppInfo> apps) {
        Collections.sort(apps, new Comparator<AppInfo>() {
            @Override public int compare(AppInfo a, AppInfo b) {
                return a.appName.toLowerCase().compareTo(b.appName.toLowerCase());
            }
        });
    }
}

File: app/src/main/java/com/azreashade/shadowpurge/data/AppRepository.kt
package com.azreashade.shadowpurge.data

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

data class AppInfo(
    val appName: String,
    val packageName: String,
    val isSystemApp: Boolean
)

class AppRepository(private val context: Context) {

    var userApps: List<AppInfo> = emptyList()
        private set

    var systemApps: List<AppInfo> = emptyList()
        private set

    suspend fun loadApps() = withContext(Dispatchers.IO) {
        val pm = context.packageManager
        val packages = pm.getInstalledApplications(PackageManager.GET_META_DATA)

        val userAppList = mutableListOf<AppInfo>()
        val systemAppList = mutableListOf<AppInfo>()

        for (app in packages) {
            val isSystem = (app.flags and ApplicationInfo.FLAG_SYSTEM) != 0
            val appInfo = AppInfo(
                appName = pm.getApplicationLabel(app).toString(),
                packageName = app.packageName,
                isSystemApp = isSystem
            )
            if (isSystem) {
                systemAppList.add(appInfo)
            } else {
                userAppList.add(appInfo)
            }
        }

        userApps = userAppList.sortedBy { it.appName.lowercase() }
        systemApps = systemAppList.sortedBy { it.appName.lowercase() }
    }
}
